tests/functional/steps/given.py

```diff
@@ -0,0 +1,177 @@
+import pytest
+from pytest_bdd import given, parsers
+from requests import post
+from tests.functional.data import searches
+from tests.functional.data import patients
+from tests.functional.data.patients import Patient
+from tests.functional.data.searches import Search
+from tests.functional.data.updates import Update
+from tests.functional.steps.when import retrieve_patient
+from pytest_nhsd_apim.apigee_apis import ApiProductsAPI
+import json
+import time
+
+
+@given('scope added to product')
+def add_scope_to_products_patient_access(products_api: ApiProductsAPI,
+                                         nhsd_apim_proxy_name: str,
+                                         nhsd_apim_authorization: dict):
+    product_name = nhsd_apim_proxy_name.replace("-asid-required", "")
+
+    default_product = products_api.get_product_by_name(product_name=product_name)
+    if nhsd_apim_authorization['scope'] not in default_product['scopes']:
+        default_product['scopes'].append(nhsd_apim_authorization['scope'])
+        products_api.put_product_by_name(product_name=product_name, body=default_product)
+        time.sleep(2)
+
+
+@given('I have a patient with a related person', target_fixture='patient')
+def patient_with_a_related_person() -> Patient:
+    return patients.WITH_RELATED_PERSON
+
+
+@given("I enter a patient's vague demographic details", target_fixture='search')
+def vague_patient() -> Search:
+    return searches.VAGUE
+
+
+@given("I am an unknown user", target_fixture='headers_with_authorization')
+def provide_headers_with_no_auth_details() -> None:
+    return {}
+
+
+@given("I am a healthcare worker")
+def provide_healthcare_worker_auth_details(request) -> None:
+    auth_details = {
+        "api_name": "personal-demographics-service",
+        "access": "healthcare_worker",
+        "level": "aal3",
+        "login_form": {"username": "656005750104"},
+        "force_new_token": True
+    }
+    request.node.add_marker(pytest.mark.nhsd_apim_authorization(auth_details))
+
+
+@given("I am a P9 user")
+def provide_p9_auth_details(request) -> None:
+    auth_details = {
+        "api_name": "personal-demographics-service",
+        "access": "patient",
+        "level": "P9",
+        "login_form": {"username": "9912003071"},
+        "force_new_token": True
+    }
```

Comment:
Would it be worth having a private method to return this (seems to be used in multiple places with different values for level and username)?

Reply:
Might have gone a little far with my amendments for this. But feels neater than just awkwardly updating `dict`s.